proptest-macro/src/property_test/utils.rs

```diff
@@ -1,41 +1,123 @@
 use core::mem::replace;
 
-use syn::{punctuated::Punctuated, FnArg, ItemFn, PatType};
+use syn::{
+    punctuated::Punctuated, AttrStyle, Attribute, Expr, FnArg, ItemFn, Meta,
+    PatType,
+};
+
+/// A parsed argument, with an optional custom strategy
+pub struct Argument {
+    pub pat_ty: PatType,
+    pub strategy: Option<Expr>,
+}
 
 /// Convert a function to a zero-arg function, and return the args
 ///
-/// Panics if any arg is a receiver (i.e. `self` or a variant)
-pub fn strip_args(mut f: ItemFn) -> (ItemFn, Vec<PatType>) {
+/// Panics on any invalid function
+pub fn strip_args(mut f: ItemFn) -> (ItemFn, Vec<Argument>) {
     let args = replace(&mut f.sig.inputs, Punctuated::new());
-    (f, args.into_iter().map(|arg| match arg {
-        FnArg::Typed(arg) => arg,
-        FnArg::Receiver(_) => panic!("receivers aren't allowed - should be filtered by `validate`"),
-    }).collect())
+    let args = args
+        .into_iter()
+        .map(|arg| match arg {
+            FnArg::Typed(arg) => strip_strategy(arg),
+            FnArg::Receiver(_) => panic!(
+                "receivers aren't allowed - should be filtered by `validate`"
+            ),
+        })
+        .collect();
+
+    (f, args)
+}
+
+fn strip_strategy(mut pat_ty: PatType) -> Argument {
+    let (strategies, others) = pat_ty.attrs.into_iter().partition(is_strategy);
+
+    pat_ty.attrs = others;
```

Comment:
attrs must always be empty here because we validate in `validate_parameter_attrs` that only `strategy = <expr>` attrs exist right? 

- if we handle the extra attrs here, why do we disallow other attrs?
- does disallowing them mean `#[property_test]` doesn't compose well with other proc macros?

Reply:
I'll double check. I'd like to compose with other macros as much as is feasible, so other attributes should be left alone. I think validate is being overly aggressive here

Reply:
But the more I think about it, I'm not sure what we could meaningfully do with the attributes, since they're going into our own struct - the output from proptest is a zero-arg function, so maybe it doesn't make sense here?

I think for now, disallowing other attributes is safer from a backwards compatibility point of view - we don't lock ourselves into supporting something that is difficult later